### DISCUSSION TITLE
Guide on generating RPM packages with fcitx for Fedora KDE

### DISCUSSION THREAD

**Discussion by ghost** [OpenBangla/OpenBangla-Keyboard#353] [General]
As OBK still does not have fcitx rpms in release tab, it has prevented users like me who intended on using Fedora Kinoite quite a headache, as the standard "make install" commands will probably not work

However, I have successfully managed to generate rpm packages with fcitx support by using/modifying commands from
https://github.com/OpenBangla/OpenBangla-Keyboard/discussions/213
https://github.com/OpenBangla/OpenBangla-Keyboard/blob/master/tools/build.sh

This guide is primarily intended to document the process.

1. You need to use some kind of container like Toolbox/Distrobox with a relatively recent Fedora image.
For toolbox it is : ```toolbox create && toolbox enter```

2. You need to install dependencies and update the image by doing the following.
```sudo dnf -y update && sudo dnf -y install @buildsys-build rust cargo cmake qt5-qtdeclarative-devel libzstd-devel fcitx5 fcitx5-devel ninja-build curl meson```

3. You need to clone the git repo, switch to develop branch and update repos
``` 
git clone --recursive https://github.com/OpenBangla/OpenBangla-Keyboard.git
cd OpenBangla-Keyboard
git checkout develop
git submodule update
mkdir build 
```

5. You can finally generate an RPM with FCITX support by executing the following command
``` cmake -Bbuild -GNinja -DCPACK_GENERATOR=RPM -DCMAKE_INSTALL_PREFIX="/usr" -DENABLE_FCITX=ON &&  ninja package -C build ```

the generated rpm will be inside the build folder

**Comment by ghost** [comment]
I have tested out the generated fcitx and the the ibus versions of OBK yesterday, KDE on Fedora 39 just won't work with any of them.

When I install the fcitx version, rpm-ostree/dnf also needs to download fcitx, so by default Fedora KDE doesn't even ship with fcitx.

After installing the fcitx version of OBK, no additional keyboard shows under the Layouts menu, even after I chose fcitx under the Virtual Keyboard menu, unlike in GNOME where OBK will show up as an input method.

So I uninstalled the fcitx OBK and installed the standard ibus OBK, set ibus Wayland as the Virtual Keyboard, then went to the Layouts menu, where an OBK layout option still was not present.

As a last-ditch effort I used the ibus-setup command in a terminal to access the ibus configuration GUI, where I could add OBK as an additional layout, it then showed up as an additional input method on the input method selector of KDE, but any attempt to switch to it would cause terrible lag and OBK would not work.

So far I donot know if this is an issue with KDE or Fedora,or maybe even a bug on OBK, as I did compile the fcitx version from the develop branch, but the ibus OBK rpm for F38 in the Release tab works flawlessly with F39 on GNOME

**Reply by mominul** [reply]
I haven't used Fedora Kinoite, so I don't know about that platform much. But after a glimpse, it seems that it might use containerization which doesn't play well with IME engines. Can you try compiling and running another fcitx IME engine to be sure that the problem is from the OBK's side?

The package installation strategy of Fedora Kinoite is different from others behind the scenes, so I think that could be a possible reason for the issue you have been facing. 

> So I uninstalled the fcitx OBK and installed the standard ibus OBK, set ibus Wayland as the Virtual Keyboard, then went to the Layouts menu, where an OBK layout option still was not present.
>
> As a last-ditch effort I used the ibus-setup command in a terminal to access the ibus configuration GUI, where I could add OBK as an additional layout, it then showed up as an additional input method on the input method selector of KDE, but any attempt to switch to it would cause terrible lag and OBK would not work.

Which version of OBK you are referencing as 'standard ibus OBK'? Version 2.0.0 or the current development version?

**Comment by ghost** [comment]
> Which version of OBK you are referencing as 'standard ibus OBK'? Version 2.0.0 or the current development version?

version 2.0.0

> it seems that it might use containerization which doesn't play well with IME engines

I installed it via rpm-ostree, rpm-ostree overlays additional packages on top of an OSTree image.
End result is that it is identical to running on actual normal Fedora itself, I have used OBK 2.0.0 on Fedora Silverblue without any issues for the past 6 months, but I will test fcitx OBK inside a standard Fedora 38 KDE VM soon.

Another thing is that in my own personal experience, KDE tends to be extremely buggy and break stuff even between minor updates, so it could be that it is just a transient issue.

I will also check under a Debian 12 KDE VM just to be extra sure.

**Comment by ghost** [comment]
![Screenshot from 2023-10-09 20-28-26](https://github.com/OpenBangla/OpenBangla-Keyboard/assets/138502262/0f0c1d20-7946-4b13-9395-26c255c2bfc6)
![Screenshot from 2023-10-09 20-27-13](https://github.com/OpenBangla/OpenBangla-Keyboard/assets/138502262/7f36da3f-b825-4e91-8782-23a342cb20b3)
![Screenshot from 2023-10-09 20-26-04](https://github.com/OpenBangla/OpenBangla-Keyboard/assets/138502262/165a143c-66fd-422e-9c23-7088da4402cb)
![Screenshot from 2023-10-09 20-25-31](https://github.com/OpenBangla/OpenBangla-Keyboard/assets/138502262/822a3bab-e880-49f5-aefc-048040d6717d)
![Screenshot from 2023-10-09 20-23-19](https://github.com/OpenBangla/OpenBangla-Keyboard/assets/138502262/a7594a05-d75f-4cbe-b638-9bba233cbe0a)

I installed Fedora 37 inside a VM, then compiled and tested OBK Fcitx inside of it.
I chose F37 specifically as KDE sometimes has tons of regressions between releases and F37 is still fairly recent and stable while also not being EoL.

As you can see, the setup process is not straightforward at all, before installing OBK Fcitx there was only None and Malit virtual keyboards.
After installing OBK Fcitx a new Fcitx option appeared, I had to select that.
After that, I tried to find OBK under the Global Layouts menu in Input > Keyboards but there was no mention of OBK, unlike GNOME where new layouts get added to the Keyboard Layouts menu in the Settings app itself

![Screenshot from 2023-10-09 20-44-54](https://github.com/OpenBangla/OpenBangla-Keyboard/assets/138502262/fd862a86-2115-46d0-8d07-9421d415a25c)

However, a new Keyboard icon then popped up in the taskbar, I had to use that to get OBK Fcitx working, it was smooth sailing after that.

Although my previous test was on Fedora Kinoite 39, knowing how KDE can be, it was probably a regression/bug of some sorts instead of an issue related to OStree, as I have been using OBK ibus on Silverblue 38/39 with zero issues.
Still I might boot up my Kinoite image again to check later.

**Reply by mominul** [reply]
Thanks for the hard work! ❤️ 

> As you can see, the setup process is not straightforward at all, before installing OBK Fcitx there was only None and Malit virtual keyboards.
After installing OBK Fcitx a new Fcitx option appeared, I had to select that.
After that, I tried to find OBK under the Global Layouts menu in Input > Keyboards but there was no mention of OBK, unlike GNOME where new layouts get added to the Keyboard Layouts menu in the Settings app itself

Yeah, the process is not smooth and straightforward, and there is very little to do from our side other than opening an issue in Fcitx/KDE project. But we need good documentation to reduce the hurdles of the users. So it would be great if you could contribute some pages regarding the installation and usage of OBK Fcitx on the [wiki](https://github.com/OpenBangla/OpenBangla-Keyboard/wiki).

Glad to know that OBK ibus is running smoothly on Fedora Silverblue, were there any special steps needed to install OBK on it?

Thanks again! 😊

**Reply by ghost** [reply]
> So it would be great if you could contribute some pages regarding the installation and usage of OBK Fcitx on the [wiki](https://github.com/OpenBangla/OpenBangla-Keyboard/wiki).

Sure, I would be glad to help out,

> Glad to know that OBK ibus is running smoothly on Fedora Silverblue, were there any special steps needed to install OBK on it?

Nope, it worked ootb without any issues.

I have installed OBK Fcitx on F39 Kinoite (KDE) on baremetal, it seems the "Keyboard" icon takes a while to pop up compared to F37.
I have used it for about 2 days on some light office works, it works well albeit with some very rare minor glitches, probably not a fault of OBK.

**Comment by ghost** [comment]
https://github.com/OpenBangla/OpenBangla-Keyboard/wiki/Configure-on-KDE-(Fcitx)

I have written a reasonably detailed wiki article on how to configure OBK Fcitx on KDE, hope this will help someone.

**Reply by ghost** [reply]
I think the main wiki article you mentioned previously https://github.com/OpenBangla/OpenBangla-Keyboard/wiki needs to be edited to include my article, I will leave it upto the OBK Team

**Reply by mominul** [reply]
Thanks a lot, I will include it in the list when version 3.0.0 is released! ❤️

**Comment by Kabir404** [comment]
If this gets merged to the master branch then the KDE or any other Qt environment installation methods should ditch IBus in the installation guide/script. As IBus doesnt really play nicely with those desktops...Only working best on GTK based desktops like GNOME or Cinnamon.

**Comment by ghost** [comment]
Just as a disclaimer, The issue I have described here: https://github.com/OpenBangla/OpenBangla-Keyboard/issues/357#issuecomment-1831867236 about Fcitx not showing up under Virtual Keyboards option in KDE Plasma 5.27.9 Wayland is a bug which happens at random.

If you install OBK Fcitx and the Fcitx option still doesn't appear under the Virtual Keyboards option, you should uninstall OBK Fcitx, reboot your computer and reinstall OBK Fcitx (preferably also do a reboot after a reinstall).
I found the fix after the bug appeared in a VM again.
This comment is to document the bug and it's fix and potentially save people some time.

Would also like to confirm that OBK Fcitx works flawlessly under KDE Plasma 6 Beta 1 on Fedora 40 Rawhide with no apparent issues in the 2-3 days I have tested KDE Plasma 6 Beta.

**Comment by ghost** [comment]
As make-pkgs has been merged, this article is now redundant, makes sense to close this.
